stl/inc/xmemory

```diff
@@ -776,6 +776,63 @@ _NODISCARD constexpr allocation_result<typename allocator_traits<_Alloc>::pointe
 // The number of user bytes a single byte of ASAN shadow memory can track.
 _INLINE_VAR constexpr size_t _Asan_granularity = 8;
 
+struct _AsanAlignedPointers {
+    const void* _First;
+    const void* _End;
+
+    _NODISCARD constexpr const void* _Clamp(const void* _Mid) const noexcept {
+        if (_Mid < _First) {
+            return _First;
+        } else if (_Mid > _End) {
+            return _End;
+        } else {
+            return _Mid;
+        }
+    }
+};
+// Returns {nullptr, nullptr} if [_First, _End) does not cover a full ASAN shadow memory section.
+// `__sanitizer_annotate_contiguous_container` _requires_
+// its first pointer to be aligned to the beginning of an ASAN shadow memory section (a single 8-byte).
+// When the `end` pointer is not on a boundary, it will also mark the remainder of the section.
+// This means that if you have, for example:
+// struct alignas(8) cat {
+//    int meow; // bytes [0, 4)
+//    char buffer[16]; // bytes [4, 20)
+//    int purr; // bytes [20, 24)
+// };
+// you cannot mark `buffer + [0, 4)`, since `buffer + 0` doesn't start a shadow memory section
+// (i.e., it's not on an 8-byte boundary).
+// Additionally, if you attempt to mark up to `buffer + 16` (the real end of buffer), it'll also mark `purr`.
+// Thus, the only part you can _actually_ mark is the complete shadow memory section of `buffer + [4, 12)`
+// (i.e., bytes [8, 16)).
+_NODISCARD inline _AsanAlignedPointers _Get_asan_aligned_first_end(

```

Comment:
@ras0219 has convinced me that we can maybe just mask off the `first` pointer without concern for being on a shadow memory boundary, since we _must_ assume that the stuff before the `first` pointer is valid...